### PR TITLE
Update docker and test configs

### PR DESCRIPTION
**Pull request by hans-thomas**
Hello there,
I updated the `Dockerfile` and `docker-compose` files. I also added some scripts to the `composer` that might be used very often.

### CODE REVIEW

File: Dockerfile
Comment (hans-thomas) [inline Dockerfile]: Minimum PHP version supported according to `composer.json` file.

File: Dockerfile
Comment (hans-thomas) [inline Dockerfile]: These lines were ineffective. I moved some of them to the `docker-compose` file. However, To run the tests, I added scripts to the `composer` to make it easy for users to run the commands.

File: .github/workflows/build-ci.yml
Comment (hans-thomas) [inline .github/workflows/build-ci.yml]: Use a composer script in `build-ci` workflow.

File: Dockerfile
Comment (GromNaN) [inline Dockerfile]: Getting the composer executable from the composer image it correct with Docker. It provides more security and allows caching.

File: .github/workflows/build-ci.yml
Comment (GromNaN) [inline .github/workflows/build-ci.yml]: Quoting strings in Yaml is a good practice to avoid unexpected behaviors like `no` being converted to `false`. So I'm not in favor of removing this quotes everywhere.

File: .github/workflows/build-ci.yml
Comment (GromNaN) [reply .github/workflows/build-ci.yml]: It makes things a little more complicated if we want to add new options to this command.

**Review by GromNaN — CHANGES_REQUESTED** [review]
Apart from composer scripts, I don't think these changes improve anything.
Thanks for the suggestion.

File: .github/workflows/build-ci.yml
Comment (hans-thomas) [reply .github/workflows/build-ci.yml]: It can be done as follows
```bash
composer test:coverage -- --new-argument
```
I don't think it's going to make things more complicated. 
Let's assume we used a command in two places with several arguments. If we want a change, we must apply that change in two places, and it increases the chance of an error.
What I suggest is, defining common commands as composer scripts and using them everywhere. If somewhere we need to add an extra argument, we can pass it like the example code I mentioned.
These scripts can make workflows cleaner and also make developing easier and faster.

**Comment by hans-thomas** [comment]
> Apart from composer scripts, I don't think these changes improve anything.

I don't want to improve performance or anything else. I just want to update the docker configuration to be more effective and make developing easier.

If you look at the current [dockerfile](https://github.com/mongodb/laravel-mongodb/blob/4.1/Dockerfile), in the last line, it is installing dependencies and run the tests. which means to run the tests, I have to build the `dockerfile` each time?
It's obvious developers must build `dockerfile` the first time and then run the tests using command line inside of the `countainer`.

File: Dockerfile
Comment (hans-thomas) [reply Dockerfile]: I replaced this line
```bash
COPY --from=composer:2.6.2 /usr/bin/composer /usr/local/bin/composer
```
Because the composer's version was fixed.
Also, we don't need to cache, because developers build this file one time, and the second reason is if they decide to rebuild this file, they definitely want the updates.

**Comment by hans-thomas** [comment]
Another update for the project's structure can be using [StyleCI](https://styleci.io/) instead of PHP Code Sniffer. StyleCI can be configured to fix the code style using an automated merge commit.
Fixing code style after every change (after the pipeline failed mostly😅) is annoying.

**Comment by GromNaN** [comment]
> If you look at the current [dockerfile](https://github.com/mongodb/laravel-mongodb/blob/4.1/Dockerfile), in the last line, it is installing dependencies and run the tests. which means to run the tests, I have to build the `dockerfile` each time? It's obvious developers must build `dockerfile` the first time and then run the tests using command line inside of the `countainer`.

The last line is the default command when you run the container.
```
docker-compose run --rm tests
```

Note that I don't use this docker image to develop because the integration with PHPStorm and xdebug set-by-step debugging is hard to setup.

So I just found that the memory limit is too low. This lines should be added after `docker-php-ext-install -j$(nproc) zip`.

```
    cp /usr/local/etc/php/php.ini-development /usr/local/etc/php/php.ini && \
    echo "memory_limit=-1" >> /usr/local/etc/php/php.ini
```

> Another update for the project's structure can be using [StyleCI](https://styleci.io/) instead of PHP Code Sniffer. StyleCI can be configured to fix the code style using an automated merge commit.
> Fixing code style after every change (after the pipeline failed mostly😅) is annoying.

Not all CS issues can be fixed automatically, but in 90% of cases, we could make it easier for external contributors to contribute by automatically committing modifications proposed by phpcbf.

**Comment by hans-thomas** [comment]
> The last line is the default command when you run the container.

It had never run tests on my computer. It is even mentioned [here](https://github.com/mongodb/laravel-mongodb/issues/2129#issuecomment-858027089).

> we could make it easier for external contributors to contribute by automatically committing modifications proposed by phpcbf.

So you mean it can be done automatically? How?

**Comment by GromNaN** [comment]
> > we could make it easier for external contributors to contribute by automatically committing modifications proposed by phpcbf.
> 
> So you mean it can be done automatically? How?

As proposed by @Treggats in [#2664](https://github.com/mongodb/laravel-mongodb/pull/2664/files#diff-dfd15ae70247bb45f121522ec2e50e5da9aed17e23116d94ec887a4769c44294).

**Comment by hans-thomas** [comment]
> As proposed by @Treggats in [#2664](https://github.com/mongodb/laravel-mongodb/pull/2664/files#diff-dfd15ae70247bb45f121522ec2e50e5da9aed17e23116d94ec887a4769c44294).

That's interesting, I might work on that PR next. Seems an important PR and the best opportunity for me to work with `PHPStan`.

**Comment by Treggats** [comment]
> As proposed by @Treggats in [#2664](https://github.com/mongodb/laravel-mongodb/pull/2664/files#diff-dfd15ae70247bb45f121522ec2e50e5da9aed17e23116d94ec887a4769c44294).

Thanks for the reminder that I needed to check the test suite. Hope that I can open it pretty soon

**Edit**
I've opened #2664 for review again.

File: Dockerfile
Comment (hans-thomas) [inline Dockerfile]: Memory limit increased.

File: Dockerfile
Comment (hans-thomas) [inline Dockerfile]: Xdebug mode is set to `coverage`.

File: composer.json
Comment (GromNaN) [inline composer.json]: We don't have `@group test` in the codebase. What's the purpose of this?

I use `./vendor/bin/phpunit --filter 'ModelTest::testNewModel'` to target a specific test (or PHPStorm PHPUnit integration of course!).

Comment (hans-thomas) [reply composer.json]: Yes, that's true. `--group` is another option to run the tests. we can select the tests that we want to execute by adding a docblock.
```php
/** @group test */
public function testWithCountHasMany(): void {}
```
The advantage is, that you can select one or more tests from several classes.
I often write multiple tests for a PR. I add this docblock to my tests to ensure that my changes don't break the previous tests. I prefer this method because adding a docblock to the code is easier (for me at least).

**Comment by hans-thomas** [comment]
I don't have any more changes to propose and I think it's the end of this PR.
Please take every feature that you find helpful and revert the rest.

**Comment by hans-thomas** [comment]
If you want to merge this PR, you can do it after merging #2664. I think there are some conflicts.

**Comment by hans-thomas** [comment]
Conflicts resolved.

File: composer.json
Comment (Treggats) [reply composer.json]: Looking through the codebase, you'd want to use the attribute for this. No docblocks are used for indicating a test or dataprovider. Well there is one, which should be a attribute in my eyes, but in general you'd want to use a attribute

```php
    #[Test]
    #[Group('hide-properties')]
    public function canHideCertainProperties(): void {}
```

More on topic of the change, personally I would not hardcode something like this. As you would use group with different groups. Not just the one defined in the composer.json file

File: Dockerfile
Comment (Treggats) [reply Dockerfile]: Is this still required after the phpunit ini setting?

File: composer.json
Comment (hans-thomas) [reply composer.json]: There is no difference between the docblock and the attribute. You can use anything you like. I just suggest adding `./vendor/bin/phpunit --group test` command as a script to save developers time to write this command each time.
Developers can use this script and add the docblock/attribute to the test cases that want to run. After that, they must remove the docblock/attribute, because that's just for testing.

> personally I would not hardcode something like this.

I'm not a big fan of hardcoding but I found this command very general and useful. As I said, It's just a suggestion, and if it doesn't make any sense, we can ignore it.

File: Dockerfile
Comment (GromNaN) [reply Dockerfile]: Please revert. You can set a less restrictive version constraint.

```php
COPY --from=composer:2 /usr/bin/composer /usr/local/bin/composer
```

> Also, we don't need to cache, because developers build this file one time, and the second reason is if they decide to rebuild this file, they definitely want the updates.

This cache is also cross-projects. It's more durable than the getcomposer.org url.

File: Dockerfile
Comment (GromNaN) [reply Dockerfile]: The memory limit in phpunit.xml is enough.
Could you merge the `RUN` stages for php.ini copy and xdebug mode.

File: composer.json
Comment (GromNaN) [reply composer.json]: Please revert.

File: .github/workflows/build-ci.yml
Comment (Treggats) [inline .github/workflows/build-ci.yml]: Should this always run, or only when debug is set?
If only when debug is set, then the following suggestion might be a bit simpler / easier to understand than the bash check.

```suggestion
            -   name: "Show PHP version"
                if: ${{ secrets.DEBUG == 'true' }}
                run: "php -v && composer -V"

```

File: Dockerfile
Comment (Treggats) [reply Dockerfile]: I'm assuming that you removed the argument because it wasn't used? Sort of inlining the version.

Which raises the question wether or not we should give the developer the option which PHP version to use.
I can imagine the tests might pass on one version and fail on the next version. But I don't work with docker on a day to day basis, so maybe my understand of this is lacking.

File: .github/workflows/build-ci.yml
Comment (Treggats) [inline .github/workflows/build-ci.yml:9]: I believe this change is based on a comment in another PR, but I wonder if it belongs in this PR. As in my eyes it is a code style thing, which adds a lot of noise to the PR. Personally I would create a separate PR to add these changes.

Though personally I don't see a benefit to these changes, but that's me and the project style is leading.

File: .github/workflows/build-ci.yml
Comment (Treggats) [inline .github/workflows/build-ci.yml:53]: Not related to the changes, as those are CS changes.

From the action-runners docs/[readme](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md)
> Both Xdebug and PCOV extensions are installed, but only Xdebug is enabled.

Does this get overridden with setting the extensions manually?

File: .github/workflows/build-ci.yml
Comment (hans-thomas) [reply .github/workflows/build-ci.yml]: The next step can be updated like this.

File: .github/workflows/build-ci.yml
Comment (hans-thomas) [reply .github/workflows/build-ci.yml]: reverted.

File: Dockerfile
Comment (hans-thomas) [reply Dockerfile]: We run the tests with the minimum supported PHP version locally. We can see the result of running tests with several versions of PHP and MongoDB when we push the changes. That's why we use strategies in the workflow.

I believe we should leave it to workflows because developers shouldn't run tests with several versions of PHP in combination with several MongoDB versions. It takes time while workflows can do it very accurately and fast.

**Comment by GromNaN** [comment]
I updated the branch with the followings:
- Revert the `PHP_VERSION` argument, so it's easy to test with other PHP versions. Even if that's not the main usage, but it would necessary to debug errors specific to a PHP version.
- Set the ini config in `phpunit.xml.dist`, it's possible to change them by creating the `phpunit.xml`
- Composer scripts automatically resolves binaries in `vendor/bin`
- Add a default command for `docker-compose run app`, that install vendors and run tests. `docker-compose run app bash` is recommanded if you need to tests many times without restarting the container.

**Comment by GromNaN** [comment]
Thank you @hans-thomas

**Pull request merged**
